File: clients/js/src/hooked/metadataDelegateRoleSeed.ts
import {
  mapSerializer,
  Serializer,
  string,
} from '@metaplex-foundation/umi/serializers';
import { TokenMetadataError } from '../errors';
import { MetadataDelegateRole } from '../generated/types/metadataDelegateRole';

export type MetadataDelegateRoleSeed =
  | 'authority_item_delegate'
  | 'collection_delegate'
  | 'use_delegate'
  | 'data_delegate'
  | 'programmable_config_delegate'
  | 'data_item_delegate'
  | 'collection_item_delegate'
  | 'prog_config_item_delegate';

export type MetadataDelegateRoleSeedArgs =
  | MetadataDelegateRoleSeed
  | MetadataDelegateRole;

export function getMetadataDelegateRoleSeedSerializer(): Serializer<
  MetadataDelegateRoleSeedArgs,
  MetadataDelegateRoleSeed
> {
  return mapSerializer(
    string({ size: 'variable' }),
    (args: MetadataDelegateRoleSeedArgs): string => {
      if (typeof args === 'string') return args;
      switch (args) {
        case MetadataDelegateRole.AuthorityItem:
          return 'authority_item_delegate';
        case MetadataDelegateRole.Collection:
          return 'collection_delegate';
        case MetadataDelegateRole.Use:
          return 'use_delegate';
        case MetadataDelegateRole.Data:
          return 'data_delegate';
        case MetadataDelegateRole.ProgrammableConfig:
          return 'programmable_config_delegate';
        case MetadataDelegateRole.DataItem:
          return 'data_item_delegate';
        case MetadataDelegateRole.CollectionItem:
          return 'collection_item_delegate';
        case MetadataDelegateRole.ProgrammableConfigItem:
          return 'prog_config_item_delegate';
        default:
          throw new TokenMetadataError(
            `Invalid MetadataDelegateRoleArgs ${args as never}`
          );
      }
    },
    (seed: string): MetadataDelegateRoleSeed => seed as MetadataDelegateRoleSeed
  );
}

File: clients/rust/src/hooked/metadata_delegate_role_seed.rs
use std::fmt::{Display, Formatter, Result};

use crate::types::MetadataDelegateRole;

pub struct MetadataDelegateRoleSeed(MetadataDelegateRole);

impl Display for MetadataDelegateRoleSeed {
    fn fmt(&self, f: &mut Formatter) -> Result {
        let message = match self.0 {
            MetadataDelegateRole::AuthorityItem => "authority_item_delegate".to_string(),
            MetadataDelegateRole::Collection => "collection_delegate".to_string(),
            MetadataDelegateRole::Use => "use_delegate".to_string(),
            MetadataDelegateRole::Data => "data_delegate".to_string(),
            MetadataDelegateRole::ProgrammableConfig => "programmable_config_delegate".to_string(),
            MetadataDelegateRole::DataItem => "data_item_delegate".to_string(),
            MetadataDelegateRole::CollectionItem => "collection_item_delegate".to_string(),
            MetadataDelegateRole::ProgrammableConfigItem => "prog_config_item_delegate".to_string(),
        };

        write!(f, "{message}")
    }
}

impl From<MetadataDelegateRole> for MetadataDelegateRoleSeed {
    fn from(role: MetadataDelegateRole) -> Self {
        MetadataDelegateRoleSeed(role)
    }
}
